File: schemas/email-campaign-marketing/convex/automation_steps.ts
// automation_steps: Ordered steps within an automation workflow (send, delay, condition).
// See README.md for full design rationale.
import { defineTable } from "convex/server";
import { v } from "convex/values";

export const automationSteps = defineTable({
  workflowId: v.id("automationWorkflows"),
  stepOrder: v.number(),
  stepType: v.union(
    v.literal("send_email"),
    v.literal("delay"),
    v.literal("condition")
  ),
  templateId: v.optional(v.id("templates")),
  config: v.optional(v.any()),
  updatedAt: v.number(),
})
  .index("by_workflow_id_step_order", ["workflowId", "stepOrder"])
  .index("by_template_id", ["templateId"]);

File: schemas/email-campaign-marketing/spacetimedb/automation_steps.rs
// automation_steps: Ordered actions within an automation workflow.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum StepType {
    // type: String
    SendEmail,
    Delay,
    Condition,
}

#[spacetimedb::table(name = automation_steps, public)]
pub struct AutomationStep {
    #[primary_key]
    pub id: String, // UUID

    pub workflow_id: String, // UUID, FK → automation_workflows.id (cascade delete)
    // Composite unique: (workflow_id, step_order) — enforce in reducer logic

    pub step_order: i32,

    pub step_type: String, // StepType enum

    #[index(btree)]
    pub template_id: Option<String>, // UUID, FK → templates.id (set null)

    pub config: Option<String>, // JSON

    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
